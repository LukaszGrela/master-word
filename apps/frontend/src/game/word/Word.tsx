import { FC, useState, useEffect, Fragment, useMemo } from 'react';
import { TValidationChar } from '@repo/backend-types';
import { classNames, isLetter } from '@repo/utils';
import { Letter } from '../cell';
import { IProps } from './types';

const Word: FC<IProps> = ({
  active,
  invalid,

  wordLength,
  word,
  id,
  commit,
  className,
  mobile,
  validated,
  language,
}) => {
  const [index, setIndex] = useState(0);
  const [guessing, setGuessing] = useState(' '.repeat(wordLength));
  const [isIncorrect, setIncorrect] = useState(invalid);

  useEffect(() => {
    setIncorrect(invalid);
  }, [invalid]);

  useEffect(() => {
    // word changed - reset
    if (word === '') {
      setIndex(0);
      setGuessing(' '.repeat(wordLength));
    }
    if (mobile && word && word !== '') {
      setIndex(word.length - 1);
      setGuessing(word);
    }
  }, [mobile, word, wordLength]);

  useEffect(() => {
    /* v8 ignore start */
    const preventDefault = (e: KeyboardEvent): void => {
      e.preventDefault();
      e.stopImmediatePropagation();
      e.stopPropagation();
    };
    /* v8 ignore stop */

    const handleKeyUp = (e: KeyboardEvent): void => {
      const action = e.key;
      if (action === 'Enter' && index === wordLength) {
        // submit word
        setIncorrect(false);
        commit(guessing);
      } else if (action === 'Backspace' && index > 0) {
        // remove letters
        const i = index - 1;
        const newGuess =
          guessing.substring(0, i) + ' ' + guessing.substring(i + 1);
        setIndex(i);
        setGuessing(newGuess);
        setIncorrect(false);
      } else if (action === 'Delete') {
        setIndex(0);
        setGuessing(' '.repeat(wordLength));
        setIncorrect(false);
      } else if (isLetter(action) && index <= wordLength) {
        // add letters
        const i = index < wordLength ? index : wordLength - 1;
        const newGuess =
          guessing.substring(0, i) +
          action.toLocaleUpperCase() +
          guessing.substring(i + 1);
        setIndex(i + 1);
        setGuessing(newGuess);
        setIncorrect(false);
      }
    };

    // listen to key up event
    if (active && !mobile) {
      document.addEventListener('keydown', preventDefault);
      document.addEventListener('keyup', handleKeyUp);
    }

    return () => {
      document.removeEventListener('keyup', handleKeyUp);
      document.removeEventListener('keydown', preventDefault);
    };
  }, [active, commit, guessing, index, mobile, wordLength]);

  const staticLetters = useMemo(() => {
    let classes = Array.from(Array(wordLength)).map(() => 'incorrect');

    if (!active && validated) {
      classes = validated.map((char: TValidationChar) => {
        if (char === 'M') {
          return 'misplaced';
        }
        if (char === 'C') {
          return 'correct';
        }
        return 'incorrect';
      });
    }

    if (word) {
      return word
        .split('')
        .map((letter, i) => (
          <Letter
            language={language}
            className={classNames(
              classes[i],
              className,
              i === 0 && 'first',
              i === wordLength - 1 && 'last',
            )}
            letter={letter}
            key={`letter-${id}-${i}`}
            data-testid={`letter-${id}-${i}`}
          />
        ));
    }

    return null;
  }, [active, className, id, validated, word, wordLength, language]);
  return (
    <Fragment>
      {active || !word
        ? guessing
            .split('')
            .map((letter, i) => (
              <Letter
                language={language}
                letter={letter}
                key={`letter-${id}-${i}`}
                data-testid={`letter-${id}-${i}`}
                className={classNames(
                  active && 'active',
                  className,
                  isIncorrect && 'wrong',
                  mobile && active && i === 0 && 'show-icon',
                  i === 0 && 'first',
                  i === wordLength - 1 && 'last',
                )}
              />
            ))
        : staticLetters}
    </Fragment>
  );
};

export default Word;
