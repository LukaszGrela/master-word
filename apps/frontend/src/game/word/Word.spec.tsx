import { fireEvent, render, waitFor } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import Word from './Word';

const hitEnter = () => {
  fireEvent.keyUp(document.activeElement || document.body, {
    key: 'Enter',
    code: 'Enter',
    charCode: 13,
  });
};
const hitDelete = () => {
  fireEvent.keyUp(document.activeElement || document.body, {
    key: 'Delete',
    code: 'Delete',
    charCode: 46,
  });
};
const hitBackspace = () => {
  fireEvent.keyUp(document.activeElement || document.body, {
    key: 'Backspace',
    code: 'Backspace',
    charCode: 8,
  });
};

describe('game', () => {
  describe('word', () => {
    describe('Word', () => {
      it('Renders', () => {
        const { container } = render(
          <Word wordLength={5} commit={vi.fn()} id="id" />,
        );

        expect(container).toMatchSnapshot();
      });

      it('Renders inactive', () => {
        const { container } = render(
          <Word
            wordLength={5}
            commit={vi.fn()}
            id="id"
            active={false}
            word="WRONG"
            validated={['C', 'M', 'X', 'X', 'X']}
            language="pl"
          />,
        );

        expect(container).toMatchSnapshot();
      });

      it('Renders active', () => {
        const { container } = render(
          <Word wordLength={5} commit={vi.fn()} id="id" active />,
        );

        expect(container).toMatchSnapshot();
      });

      it('Renders active mobile', () => {
        const { container } = render(
          <Word wordLength={5} commit={vi.fn()} id="id" active mobile />,
        );

        expect(container).toMatchSnapshot();
      });
      it('Renders invalid', () => {
        const { container } = render(
          <Word
            wordLength={5}
            commit={vi.fn()}
            id="id"
            active
            invalid
            word="WRONG"
          />,
        );

        expect(container).toMatchSnapshot();
      });

      it('Renders mobile, update word', () => {
        const { rerender, getByTestId } = render(
          <Word
            wordLength={5}
            commit={vi.fn()}
            id="mobile"
            active
            mobile
            word=""
          />,
        );

        let firstLetter = getByTestId('letter-mobile-0');
        expect(firstLetter).toBeDefined();

        // empty string
        expect(firstLetter.textContent).toBe(' ');

        rerender(
          <Word
            wordLength={5}
            commit={vi.fn()}
            id="mobile"
            active
            mobile
            word="WRONG"
          />,
        );

        firstLetter = getByTestId('letter-mobile-0');
        expect(firstLetter).toBeDefined();

        // empty string
        expect(firstLetter.textContent).toBe('W');
      });

      it('If mobile or not active, no keyboard events', async () => {
        const commit = vi.fn();
        const { rerender, getByTestId } = render(
          <Word
            wordLength={5}
            commit={commit}
            id="mobile"
            active
            mobile
            word=""
          />,
        );

        let firstLetter = getByTestId('letter-mobile-0');
        expect(firstLetter).toBeDefined();

        // empty string
        expect(firstLetter.textContent).toBe(' ');

        // Attempt to type word
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 'g',
          code: 'KeyG',
          charCode: 71,
        });

        firstLetter = getByTestId('letter-mobile-0');
        expect(firstLetter).toBeDefined();

        // still empty string
        expect(firstLetter.textContent).toBe(' ');

        rerender(
          <Word
            wordLength={5}
            commit={commit}
            id="mobile"
            active
            mobile={false}
            word=""
          />,
        );

        // Attempt to type word GUESS
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 'g',
          code: 'KeyG',
          charCode: 71,
        });
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 'u',
          code: 'KeyU',
          charCode: 85,
        });
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 'e',
          code: 'KeyE',
          charCode: 69,
        });
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 's',
          code: 'KeyS',
          charCode: 83,
        });
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 's',
          code: 'KeyS',
          charCode: 83,
        });

        firstLetter = getByTestId('letter-mobile-0');
        expect(firstLetter).toBeDefined();

        //
        expect(firstLetter.textContent).toBe('G');

        // event
        hitEnter();

        await waitFor(() => {
          expect(commit).toHaveBeenLastCalledWith('GUESS');
        });
      });
      it('Test Backspace and Delete', () => {
        const commit = vi.fn();
        const { getByTestId } = render(
          <Word wordLength={5} commit={commit} id="mobile" active word="" />,
        );

        let firstLetter = getByTestId('letter-mobile-0');
        expect(firstLetter).toBeDefined();

        // empty string
        expect(firstLetter.textContent).toBe(' ');

        // Attempt to type word GUESS
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 'g',
          code: 'KeyG',
          charCode: 71,
        });
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 'u',
          code: 'KeyU',
          charCode: 85,
        });
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 'e',
          code: 'KeyE',
          charCode: 69,
        });
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 's',
          code: 'KeyS',
          charCode: 83,
        });
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 's',
          code: 'KeyS',
          charCode: 83,
        });

        let lastLetter = getByTestId('letter-mobile-4');
        expect(lastLetter).toBeDefined();
        expect(lastLetter.textContent).toBe('S');

        // overflow strokes will replace last letter
        fireEvent.keyUp(document.activeElement || document.body, {
          key: 't',
          code: 'KeyT',
          charCode: 84,
        });

        lastLetter = getByTestId('letter-mobile-4');
        expect(lastLetter).toBeDefined();
        expect(lastLetter.textContent).toBe('T');

        // event
        hitBackspace();

        firstLetter = getByTestId('letter-mobile-0');
        expect(firstLetter).toBeDefined();
        expect(firstLetter.textContent).toBe('G');
        lastLetter = getByTestId('letter-mobile-4');
        expect(lastLetter).toBeDefined();
        expect(lastLetter.textContent).toBe(' ');

        // event
        hitDelete();

        firstLetter = getByTestId('letter-mobile-0');
        expect(firstLetter).toBeDefined();
        expect(firstLetter.textContent).toBe(' ');
        lastLetter = getByTestId('letter-mobile-4');
        expect(lastLetter).toBeDefined();
        expect(lastLetter.textContent).toBe(' ');
      });
    });
  });
});
