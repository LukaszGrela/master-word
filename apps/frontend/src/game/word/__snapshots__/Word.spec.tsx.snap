// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`game > word > Word > Renders 1`] = `
<div>
  <div
    class="letter first"
    data-testid="letter-id-0"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter"
    data-testid="letter-id-1"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter"
    data-testid="letter-id-2"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter"
    data-testid="letter-id-3"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter last"
    data-testid="letter-id-4"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
</div>
`;

exports[`game > word > Word > Renders active 1`] = `
<div>
  <div
    class="letter active first"
    data-testid="letter-id-0"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter active"
    data-testid="letter-id-1"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter active"
    data-testid="letter-id-2"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter active"
    data-testid="letter-id-3"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter active last"
    data-testid="letter-id-4"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
</div>
`;

exports[`game > word > Word > Renders active mobile 1`] = `
<div>
  <div
    class="letter active show-icon first"
    data-testid="letter-id-0"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter active"
    data-testid="letter-id-1"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter active"
    data-testid="letter-id-2"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter active"
    data-testid="letter-id-3"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter active last"
    data-testid="letter-id-4"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
</div>
`;

exports[`game > word > Word > Renders inactive 1`] = `
<div>
  <div
    class="letter correct first"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
      W
    </span>
  </div>
  <div
    class="letter misplaced"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
      R
    </span>
  </div>
  <div
    class="letter incorrect"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
      O
    </span>
  </div>
  <div
    class="letter incorrect"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
      N
    </span>
  </div>
  <div
    class="letter incorrect last"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
      G
    </span>
  </div>
</div>
`;

exports[`game > word > Word > Renders invalid 1`] = `
<div>
  <div
    class="letter active wrong first"
    data-testid="letter-id-0"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter active wrong"
    data-testid="letter-id-1"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter active wrong"
    data-testid="letter-id-2"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter active wrong"
    data-testid="letter-id-3"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
  <div
    class="letter active wrong last"
    data-testid="letter-id-4"
  >
    <span
      class="flag"
    >
      <span
        class="flag-wrapper"
      >
        <img
          alt=""
          loading="lazy"
          src="https://flagcdn.com/w20/pl.png"
          srcset="https://flagcdn.com/w40/pl.png 2x"
          width="20"
        />
      </span>
    </span>
    <span
      class="icon"
    >
      <svg
        class="edit-icon"
        height="48"
        viewBox="0 -960 960 960"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M167-120q-21 5-36.5-10.5T120-167l40-191 198 198-191 40Zm191-40L160-358l473-473q17-17 42-17t42 17l114 114q17 17 17 42t-17 42L358-160Zm317-628L233-346l113 113 442-442-113-113Z"
        />
      </svg>
    </span>
    <span
      class="container"
    >
       
    </span>
  </div>
</div>
`;
